docs/OBJECT-LEVEL-ENCRYPTION.md

```diff
@@ -0,0 +1,119 @@
+# OBJECT LEVEL ENCRYPTION
+
+## Motivation
+
+Independent of the use of TLS in the [API](../api/APIv2.md) and [object signing](../api/OBJECT-SIGNING.md) there are some secrets in the device configuration API (such as datastore credentials, WiFi credentials, and cloud-init user data) which should have minimal exposure by
+
+1. keeping them in some vault inside a controller implementation
+2. not exposing them to any microservices in a controller except for the party which decrypts from the vault and re-encrypts for a particular device
+3. not exposing them to all microservices on the device and accidentally storing them in files or putting them in logs
+4. potentially restricting it by policy on the device so that in a defense-in-depth implementation only the downloader can access datastore credentials, only nim can access WiFi credentials (basically least privilege approach)
+5. MiTM proxies and server-side load balancers which terminate TLS should not be able to see this information
+
+## Overview
+
+The approach is to encrypt such credential and other private information using a per-device key, and doing this encryption before assigning it to the fields which get protobuf encoded. Thus other microservices in a controller and the device will pass them along as any protobuf to golang data structure, but the content will be encrypted.
+
+This means that the protobuf API will have new fields which are the encrypted fields (with the old username/password fields etc still in place as we transition away from only protecting this using TLS).
```

Comment:
Is it necessary? Can we get away with building the assumption into APIv2 that these fields are encrypted? And if some services fail, well, they are old API anyways?

Reply:
v2API has been running for a few months already. And we've tested this object encryption on top of the v2API. But we didn't join the schedules at the hip before we started the v2API work.

Note that in general we introduce a new protobuf field and add some comment about "deprecate" for the old one, and get around to that later once the controllers and EVE have all moved to the new field.

Reply:
Got it.